src/engine/GroupBy.cpp

```diff
@@ -404,6 +413,128 @@ ProtoResult GroupBy::computeResult([[maybe_unused]] bool requestLaziness) {
+    std::vector<HashMapAliasInformation> aggregateAliases,
+    std::vector<size_t> groupByCols, std::shared_ptr<LocalVocab> localVocab,
+    bool singleIdTable) const {
+  AD_CONTRACT_CHECK(groupByCols.size() == NUM_GROUP_COLUMNS);
+  std::vector<const LocalVocab*> vocabs{localVocab.get(),
+                                        &subresult->localVocab()};
+
+  LazyGroupBy<NUM_GROUP_COLUMNS> lazyGroupBy{
+      *localVocab, std::move(aggregateAliases), groupByCols,
+      getExecutionContext()->getAllocator()};
+
+  auto createEvaluationContextForTable = [this, &localVocab](IdTable& table) {
+    sparqlExpression::EvaluationContext evaluationContext{
+        *getExecutionContext(),
+        _subtree->getVariableColumns(),
+        table,
+        getExecutionContext()->getAllocator(),
+        *localVocab,
+        cancellationHandle_,
+        deadline_};
+    evaluationContext._groupedVariables = ad_utility::HashSet<Variable>{
+        _groupByVariables.begin(), _groupByVariables.end()};
+    evaluationContext._isPartOfGroupBy = true;
+    evaluationContext._variableToColumnMapPreviousResults =
+        getInternallyVisibleVariableColumns();
+    evaluationContext._previousResultsFromSameGroup.resize(getResultWidth());
+    return evaluationContext;
+  };
+
+  IdTable resultTable{getResultWidth(), getExecutionContext()->getAllocator()};
+
+  bool groupSplitAcrossTables = false;
+
+  for (IdTable& idTable : subresult->idTables()) {
+    if (idTable.empty()) {
+      continue;
+    }
+    checkCancellation();
+    *localVocab = LocalVocab::merge(vocabs);
+
+    lazyGroupBy.populateGroupBlock(idTable, 0);
+
+    sparqlExpression::EvaluationContext evaluationContext =
+        createEvaluationContextForTable(idTable);
+
+    size_t blockStart = 0;
+
+    for (size_t pos = 0; pos < idTable.size(); pos++) {
+      checkCancellation();
+      if (!lazyGroupBy.rowMatchesCurrentBlock(idTable, pos)) {
+        if (groupSplitAcrossTables) {
+          lazyGroupBy.processNextBlock(evaluationContext, blockStart, pos);
+          lazyGroupBy.commitRow(resultTable, evaluationContext, *this);
+          groupSplitAcrossTables = false;
+        } else {
+          // This processes the whole block in batches if possible
+          resultTable.emplace_back();
+          for (size_t i = 0; i < groupByCols.size(); ++i) {
+            resultTable(resultTable.size() - 1, i) =
+                idTable(blockStart, groupByCols[i]);
+          }
+          IdTableStatic<OUT_WIDTH> table =
+              std::move(resultTable).toStatic<OUT_WIDTH>();
+          for (const Aggregate& aggregate : aggregates) {
+            processGroup<OUT_WIDTH>(aggregate, evaluationContext, blockStart,
+                                    pos, &table, table.size() - 1,
+                                    aggregate._outCol, localVocab.get());
+          }
+          resultTable = std::move(table).toDynamic();
```

Comment:
This code is a duplication of something inside
`doGroupBy`.  This should at least be refactored.
And I still would argue for a function
`processAllGroupsBetween(startOfFirstGroup, endOfLastGroups, IdTable* result)` 
that can be shared for the materialized group by (then the first two arguments are 0 and input.size()), 
as well as the lazy GROUP BY, where we leave out the first and the last group of a block, because they might span across multiple blocks and need to be handled by the `lazyGroupBy` class.